src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -110,12 +118,62 @@ protected Object readResolve() throws ObjectStreamException {
      * It will also be false for something that has finished but is pending child node creation,
      * such as a completed fork branch which is waiting for the join node to be created.
      * <p>This can only go from true to false and is a shortcut for {@link FlowExecution#isCurrentHead}.
+     * @see #isActive
      */
     @Exported
     public final boolean isRunning() {
         return getExecution().isCurrentHead(this);
     }
 
+    private static final Map<FlowExecution, Set<BlockStartNode>> unclosedBlocks = new WeakHashMap<>();
```

Comment:
So, how will this behave if the master restarts?    Normally the FlowExecution would be safe to use as a weak key because we only care about running flows, where a strong reference will be held.  But won't this be unpopulated when we restart, meaning we can't refer to it for information on running flows?